README.md
<p align=center>
  <a href="https://github.com/TimoKunze/ButtonControls/releases/tag/1.10.4">
    <img alt="Release 1.10.4 Release" src="https://img.shields.io/badge/release-1.10.4-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/ButtonControls/releases">
    <img alt="Download ButtonControls" src="https://img.shields.io/badge/download-latest-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/ButtonControls/blob/master/LICENSE">
    <img alt="License" src="https://img.shields.io/badge/license-MIT-0688CB.svg">
  </a>
  <a href="https://www.paypal.com/xclick/business=TKunze71216%40gmx.de&item_name=ButtonControls&no_shipping=1&tax=0&currency_code=EUR">
    <img alt="Donate" src="https://img.shields.io/badge/%24-donate-E44E4A.svg">
  </a>
</p>

# ButtonControls
An ActiveX control library for Visual Basic 6 that contains various kinds of buttons, check boxes, radio buttons and frames.

I've developed this ActiveX control in 2005 and did update it on a regular basis until 2016. Currently I have little interest to maintain this project any longer, but I think the code might be of some use to others.

# Before you make changes
If you make changes to the code and deploy the binary, keep in mind that ActiveX controls are COM components and therefore should stay binary compatible as long as you don't change the COM object's, i.e. the ActiveX control's public class name and GUIDs. Otherwise people using these components are likely to end up in the famous COM hell.

# Dependencies
You will need the [Microsoft Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk), ATL, and [WTL 10](https://sourceforge.net/projects/wtl/).

## Required Change to ATL
Some version of ATL have a bug in ```AtlIPersistPropertyBag_Load``` which causes crashes. In the file atlcom.h search for ```AtlIPersistPropertyBag_Load```. Inside the implementation of this function search for ```HRESULT hr = pPropBag->Read(pMap[i].szDesc, &var, pErrorLog);```. Make sure that there is this code before this line: ```var.pdispVal = NULL;```.

CHANGELOG
Version 1.10.4 - 21/01/2018

- Upgraded to WTL 10.
- Integrated the GitHub repository into the documentation and the About dialog.



Version 1.10.3 - 07/01/2018

- Project published on GitHub.



Version 1.10.2 - 30/07/2016

- Fixed some memory leaks that occurred when firing certain events.



Version 1.10.1 - 28/01/2016

- FIX: The controls could crash during OLE drag'n'drop.
- FIX: The controls could crash the other process during OLE drag'n'drop, especially when IOLEDataObject::SetDropDescription() had been called.



Version 1.10.0 - 19/01/2016

- Added the FinishOLEDragDrop method to each control. It should not be used if not really needed. It can be used to hide the drag image if execution of the OLEDragDrop event takes a long time or if a message is displayed from within this event.
- Initial support for Windows 10.



Version 1.9.0 - 29/05/2014

- Added the XClick, and XDblClick events for support of the extended mouse buttons.
- Extended the Mouse* events to support the extended mouse buttons.
- FIX: The Font and MouseIcon properties could leak GDI resources.
- Initial support for Windows 8.1.



Version 1.8.9 - 17/10/2013

- Reactivated data-binding capabilities. When the library is registered, it will call regtlib.exe msdatsrc.tlb before registering itself. This seems to fix the mess caused by the broken Windows Update from the beginning of this year.



Version 1.8.8 - 06/10/2013

- Slightly reduced GDI resource usage.



Version 1.8.7 - 29/03/2013

- Disabled data-binding capabilities as they cause "Object library not registered" errors when adding the control library to the project. This problem might be related to a recent Windows Update.



Version 1.8.6 - 09/03/2013

- FIX: The Frame control has not been drawn correctly if the RightToLeft property included the rtlLayout flag.



Version 1.8.5 - 13/01/2013

- FIX: The GetIdealSize method ignored the Appearance and BorderStyle properties.
- FIX: Drawing issues could occur if a windowless control was placed inside the Frame control.



Version 1.8.4 - 14/10/2012

- FIX: The Frame control wasn't redrawn properly if its text has been shortened.
- FIX: The last character of the control text could not be used as accelerator key.
- FIX: The ResizedControlWindow event was not fired when it should have been fired.



Version 1.8.3 - 08/09/2012

- Fixed an error in the Frame control's image list handling code that could lead to crashes.
- FIX: Windowless controls placed inside the Frame control did not raise Mouse events anymore.
- Initial support for Windows 8.



Version 1.8.2 - 15/04/2012

- FIX: The Frame control could receive the keyboard focus by double-clicking on it.
- FIX: The Frame control could make child controls lose the keyboard focus when they shouldn't. This problem occured in MDI applications only.



Version 1.8.1 - 11/03/2012

- Further performance improvements.



Version 1.8.0 - 10/03/2012

- Rewrote the drawing code again, because the previous version did not perform well.
- Reimplemented the subclassing code of the VB6 samples to make it more robust.



Version 1.7.0 - 08/01/2012

- Rewrote the drawing code of all controls for the case that the client application uses version 6.x of comctl32.dll. This should fix drawing issues, especially with the Frame control.



Version 1.6.1 - 20/05/2011

- Fixed a bug in the Frame control that could make other, invisible windows visible again under rather complex circumstances.
- FIX: Disabling the controls did not move the keyboard focus to the next control in the tab order.
- FIX: The ResizedControlWindow event was fired when actually nothing had changed.



Version 1.6.0 - 07/01/2011

- FIX: Due do a bug in Microsoft ATL, the controls could lose keyboard focus when they shouldn't. This problem occured in MDI applications only. Many thanks to Peter Schellenbach for his help.
- Switched to Visual C++ 2010 and dropped support for Windows 2000.



Version 1.5.1 - 07/11/2010

- FIX: IOLEDataObject::SetDropDescription could cause crashes in other programs.
- Added a new VB6 sample: Color Chooser.



Version 1.5.0 - 02/05/2010

- Added the IconIndex and UseImprovedImageListSupport properties to each control. These properties make it possible to use a single image list for all button controls on a form or in an app.
- Added a new property page that can be used to enter Unicode texts at design time.



Version 1.4.3 - 01/01/2010

- Improved performance.
- FIX: If the controls were placed on a Form for which the KeyPreview feature was activated, the Form_Key* handler didn't receive key strokes like Delete, if the controls had the keyboard focus.
- FIX: The 'shift' parameter that many events have, often missed the ALT key.
- Deploy (stripped) symbols with the binaries to make debugging easier. This are the pdb files in case you wonder.



Version 1.4.2 - 03/10/2009

- Changed the implementation of the ICheckBox::SelectionState property, so that setting it to -1 is equal to setting it to ssChecked. This is useful for data-binding.
- Improved support for Aero drag images.



Version 1.4.1 - 25/07/2009

- Massively reduced the usage of GDI handles.
- Made some internal changes that will make it easier for me to keep binary compatibility in future.



Version 1.4.0 - 10/04/2009

- Initial support for Windows 7.
- Extended the DropDescriptionIconConstants enumeration by 'ddiUseDefault'.
- FIX: Drop descriptions didn't work as expected if the drag source was a 64 bit app.
- FIX: Changing the RightToLeft properties at runtime did not always have the expected effect.
- Improved performance of Frame control.



Version 1.3.1 - 29/10/2008

- FIX: Ignore calls to IOLEDataObject::SetDropDescription() if the drag image is in classic style.
- Slightly changed handling of the situation where the user presses a mouse button, drags the mouse out of the control's client area and releases the button.
- Fixed some GDI handle leaks.
- Fixed some drawing glitches.

src/DispIDs.h
// DispIDs.h: Defines a DISPID for each COM property and method we're providing.

// ICheckBox
// properties
#define DISPID_CHKBOX_APPEARANCE										1
#define DISPID_CHKBOX_AUTOTOGGLECHECKMARK						2
#define DISPID_CHKBOX_BACKCOLOR											3
#define DISPID_CHKBOX_BORDERSTYLE										4
#define DISPID_CHKBOX_CHECKMARKONRIGHT							5
#define DISPID_CHKBOX_CONTENTTYPE										6
#define DISPID_CHKBOX_DISABLEDEVENTS								7
#define DISPID_CHKBOX_DONTREDRAW										8
#define DISPID_CHKBOX_ENABLED												DISPID_ENABLED
#define DISPID_CHKBOX_FONT													9
#define DISPID_CHKBOX_FORECOLOR											10
#define DISPID_CHKBOX_HALIGNMENT										11
#define DISPID_CHKBOX_HIMAGELIST										12
#define DISPID_CHKBOX_HOVERTIME											13
#define DISPID_CHKBOX_HWND													14
#define DISPID_CHKBOX_ICONALIGNMENT									15
#define DISPID_CHKBOX_ICONMARGINBOTTOM							16
#define DISPID_CHKBOX_ICONMARGINLEFT								17
#define DISPID_CHKBOX_ICONMARGINRIGHT								18
#define DISPID_CHKBOX_ICONMARGINTOP									19
#define DISPID_CHKBOX_IMAGE													20
#define DISPID_CHKBOX_MOUSEICON											21
#define DISPID_CHKBOX_MOUSEPOINTER									22
#define DISPID_CHKBOX_MULTILINE											23
#define DISPID_CHKBOX_PROCESSCONTEXTMENUKEYS				24
#define DISPID_CHKBOX_PUSHED												25
#define DISPID_CHKBOX_PUSHLIKE											26
#define DISPID_CHKBOX_REGISTERFOROLEDRAGDROP				27
#define DISPID_CHKBOX_RIGHTTOLEFT										28
#define DISPID_CHKBOX_SELECTIONSTATE								29
#define DISPID_CHKBOX_STYLE													30
#define DISPID_CHKBOX_SUPPORTOLEDRAGIMAGES					31
#define DISPID_CHKBOX_TEXT													32
#define DISPID_CHKBOX_TEXTMARGINBOTTOM							33
#define DISPID_CHKBOX_TEXTMARGINLEFT								34
#define DISPID_CHKBOX_TEXTMARGINRIGHT								35
#define DISPID_CHKBOX_TEXTMARGINTOP									36
#define DISPID_CHKBOX_TRISTATE											37
#define DISPID_CHKBOX_USESYSTEMFONT									38
#define DISPID_CHKBOX_VALIGNMENT										39
#define DISPID_CHKBOX_VERSION												40
#define DISPID_CHKBOX_ICONINDEX											46
#define DISPID_CHKBOX_USEIMPROVEDIMAGELISTSUPPORT		47
// fingerprint
#define DISPID_CHKBOX_APPID													500
#define DISPID_CHKBOX_APPNAME												501
#define DISPID_CHKBOX_APPSHORTNAME									502
#define DISPID_CHKBOX_BUILD													503
#define DISPID_CHKBOX_CHARSET												504
#define DISPID_CHKBOX_ISRELEASE											505
#define DISPID_CHKBOX_PROGRAMMER										506
#define DISPID_CHKBOX_TESTER												507
// methods
#define DISPID_CHKBOX_ABOUT													DISPID_ABOUTBOX
#define DISPID_CHKBOX_CLICK													41
#define DISPID_CHKBOX_GETCONTROLSTATE								42
#define DISPID_CHKBOX_LOADSETTINGSFROMFILE					43
#define DISPID_CHKBOX_REFRESH												DISPID_REFRESH
#define DISPID_CHKBOX_SAVESETTINGSTOFILE						44
#define DISPID_CHKBOX_GETIDEALSIZE									45
#define DISPID_CHKBOX_FINISHOLEDRAGDROP							48


// _ICheckBoxEvents
// methods
#define DISPID_CHKBOXE_CLICK												1
#define DISPID_CHKBOXE_CONTEXTMENU									2
#define DISPID_CHKBOXE_CUSTOMDRAW										3
#define DISPID_CHKBOXE_DBLCLICK											4
#define DISPID_CHKBOXE_DESTROYEDCONTROLWINDOW				5
#define DISPID_CHKBOXE_KEYDOWN											6
#define DISPID_CHKBOXE_KEYPRESS											7
#define DISPID_CHKBOXE_KEYUP												8
#define DISPID_CHKBOXE_MCLICK												9
#define DISPID_CHKBOXE_MDBLCLICK										10
#define DISPID_CHKBOXE_MOUSEDOWN										11
#define DISPID_CHKBOXE_MOUSEENTER										12
#define DISPID_CHKBOXE_MOUSEHOVER										13
#define DISPID_CHKBOXE_MOUSELEAVE										14
#define DISPID_CHKBOXE_MOUSEMOVE										15
#define DISPID_CHKBOXE_MOUSEUP											16
#define DISPID_CHKBOXE_OLEDRAGDROP									17
#define DISPID_CHKBOXE_OLEDRAGENTER									18
#define DISPID_CHKBOXE_OLEDRAGLEAVE									19
#define DISPID_CHKBOXE_OLEDRAGMOUSEMOVE							20
#define DISPID_CHKBOXE_OWNERDRAW										21
#define DISPID_CHKBOXE_RCLICK												22
#define DISPID_CHKBOXE_RDBLCLICK										23
#define DISPID_CHKBOXE_RECREATEDCONTROLWINDOW				24
#define DISPID_CHKBOXE_RESIZEDCONTROLWINDOW					25
#define DISPID_CHKBOXE_SELECTIONSTATECHANGED				DISPID_VALUE
#define DISPID_CHKBOXE_XCLICK												26
#define DISPID_CHKBOXE_XDBLCLICK										27


// ICommandButton
// properties
#define DISPID_CMDBTN_APPEARANCE										1
#define DISPID_CMDBTN_BACKCOLOR											2
#define DISPID_CMDBTN_BORDERSTYLE										3
#define DISPID_CMDBTN_CONTENTTYPE										4
#define DISPID_CMDBTN_DISABLEDEVENTS								5
#define DISPID_CMDBTN_DONTREDRAW										6
#define DISPID_CMDBTN_ENABLED												DISPID_ENABLED
#define DISPID_CMDBTN_FONT													7
#define DISPID_CMDBTN_FORECOLOR											8
#define DISPID_CMDBTN_HALIGNMENT										9
#define DISPID_CMDBTN_HIMAGELIST										10
#define DISPID_CMDBTN_HOVERTIME											11
#define DISPID_CMDBTN_HWND													12
#define DISPID_CMDBTN_ICONALIGNMENT									13
#define DISPID_CMDBTN_ICONMARGINBOTTOM							14
#define DISPID_CMDBTN_ICONMARGINLEFT								15
#define DISPID_CMDBTN_ICONMARGINRIGHT								16
#define DISPID_CMDBTN_ICONMARGINTOP									17
#define DISPID_CMDBTN_IMAGE													18
#define DISPID_CMDBTN_MOUSEICON											19
#define DISPID_CMDBTN_MOUSEPOINTER									20
#define DISPID_CMDBTN_MULTILINE											21
#define DISPID_CMDBTN_PROCESSCONTEXTMENUKEYS				22
#define DISPID_CMDBTN_PUSHED												23
#define DISPID_CMDBTN_REGISTERFOROLEDRAGDROP				24
#define DISPID_CMDBTN_RIGHTTOLEFT										25
#define DISPID_CMDBTN_STYLE													26
#define DISPID_CMDBTN_SUPPORTOLEDRAGIMAGES					27
#define DISPID_CMDBTN_TEXT													28
#define DISPID_CMDBTN_TEXTMARGINBOTTOM							29
#define DISPID_CMDBTN_TEXTMARGINLEFT								30
#define DISPID_CMDBTN_TEXTMARGINRIGHT								31
#define DISPID_CMDBTN_TEXTMARGINTOP									32
#define DISPID_CMDBTN_USESYSTEMFONT									33
#define DISPID_CMDBTN_VALIGNMENT										34
#define DISPID_CMDBTN_VERSION												35
#define DISPID_CMDBTN_BUTTONTYPE										41
#define DISPID_CMDBTN_COMMANDLINKNOTE								42
#define DISPID_CMDBTN_DROPDOWNARROWHEIGHT						43
#define DISPID_CMDBTN_DROPDOWNARROWWIDTH						44
#define DISPID_CMDBTN_DROPDOWNGLYPH									45
#define DISPID_CMDBTN_DROPDOWNONRIGHT								46
#define DISPID_CMDBTN_DROPDOWNPUSHED								47
#define DISPID_CMDBTN_DROPDOWNSTYLE									48
#define DISPID_CMDBTN_HDROPDOWNIMAGELIST						49
#define DISPID_CMDBTN_KEEPDROPDOWNARROWASPECTRATIO	50
#define DISPID_CMDBTN_SHOWRIGHTSELEVATIONICON				51
#define DISPID_CMDBTN_SHOWSPLITLINE									52
#define DISPID_CMDBTN_ICONINDEX											53
#define DISPID_CMDBTN_USEIMPROVEDIMAGELISTSUPPORT		54
// fingerprint
#define DISPID_CMDBTN_APPID													500
#define DISPID_CMDBTN_APPNAME												501
#define DISPID_CMDBTN_APPSHORTNAME									502
#define DISPID_CMDBTN_BUILD													503
#define DISPID_CMDBTN_CHARSET												504
#define DISPID_CMDBTN_ISRELEASE											505
#define DISPID_CMDBTN_PROGRAMMER										506
#define DISPID_CMDBTN_TESTER												507
// methods
#define DISPID_CMDBTN_ABOUT													DISPID_ABOUTBOX
#define DISPID_CMDBTN_CLICK													36
#define DISPID_CMDBTN_GETCONTROLSTATE								37
#define DISPID_CMDBTN_GETIDEALSIZE									38
#define DISPID_CMDBTN_LOADSETTINGSFROMFILE					39
#define DISPID_CMDBTN_REFRESH												DISPID_REFRESH
#define DISPID_CMDBTN_SAVESETTINGSTOFILE						40
#define DISPID_CMDBTN_FINISHOLEDRAGDROP							55


// _ICommandButtonEvents
// methods
#define DISPID_CMDBTNE_CLICK												DISPID_VALUE
#define DISPID_CMDBTNE_CONTEXTMENU									1
#define DISPID_CMDBTNE_CUSTOMDRAW										2
#define DISPID_CMDBTNE_DBLCLICK											3
#define DISPID_CMDBTNE_DESTROYEDCONTROLWINDOW				4
#define DISPID_CMDBTNE_KEYDOWN											5
#define DISPID_CMDBTNE_KEYPRESS											6
#define DISPID_CMDBTNE_KEYUP												7
#define DISPID_CMDBTNE_MCLICK												8
#define DISPID_CMDBTNE_MDBLCLICK										9
#define DISPID_CMDBTNE_MOUSEDOWN										10
#define DISPID_CMDBTNE_MOUSEENTER										11
#define DISPID_CMDBTNE_MOUSEHOVER										12
#define DISPID_CMDBTNE_MOUSELEAVE										13
#define DISPID_CMDBTNE_MOUSEMOVE										14
#define DISPID_CMDBTNE_MOUSEUP											15
#define DISPID_CMDBTNE_OLEDRAGDROP									16
#define DISPID_CMDBTNE_OLEDRAGENTER									17
#define DISPID_CMDBTNE_OLEDRAGLEAVE									18
#define DISPID_CMDBTNE_OLEDRAGMOUSEMOVE							19
#define DISPID_CMDBTNE_OWNERDRAW										20
#define DISPID_CMDBTNE_RCLICK												21
#define DISPID_CMDBTNE_RDBLCLICK										22
#define DISPID_CMDBTNE_RECREATEDCONTROLWINDOW				23
#define DISPID_CMDBTNE_RESIZEDCONTROLWINDOW					24
#define DISPID_CMDBTNE_CUSTOMDROPDOWNAREASIZE				25
#define DISPID_CMDBTNE_DROPDOWN											26
#define DISPID_CMDBTNE_XCLICK												27
#define DISPID_CMDBTNE_XDBLCLICK										28


// IFrame
// properties
#define DISPID_FRM_APPEARANCE												1
#define DISPID_FRM_BACKCOLOR												2
#define DISPID_FRM_BORDERSTYLE											3
#define DISPID_FRM_BORDERVISIBLE										4
#define DISPID_FRM_CONTENTTYPE											5
#define DISPID_FRM_DISABLEDEVENTS										6
#define DISPID_FRM_DONTREDRAW												7
#define DISPID_FRM_ENABLED													DISPID_ENABLED
#define DISPID_FRM_FONT															8
#define DISPID_FRM_FORECOLOR												9
#define DISPID_FRM_HALIGNMENT												10
#define DISPID_FRM_HIMAGELIST												11
#define DISPID_FRM_HOVERTIME												12
#define DISPID_FRM_HWND															13
#define DISPID_FRM_ICONALIGNMENT										14
#define DISPID_FRM_ICONMARGINBOTTOM									15
#define DISPID_FRM_ICONMARGINLEFT										16
#define DISPID_FRM_ICONMARGINRIGHT									17
#define DISPID_FRM_ICONMARGINTOP										18
#define DISPID_FRM_IMAGE														19
#define DISPID_FRM_MOUSEICON												20
#define DISPID_FRM_MOUSEPOINTER											21
#define DISPID_FRM_REGISTERFOROLEDRAGDROP						22
#define DISPID_FRM_RIGHTTOLEFT											23
#define DISPID_FRM_STYLE														24
#define DISPID_FRM_SUPPORTOLEDRAGIMAGES							25
#define DISPID_FRM_TEXT															26
#define DISPID_FRM_USESYSTEMFONT										27
#define DISPID_FRM_VERSION													28
#define DISPID_FRM_ICONINDEX												33
#define DISPID_FRM_USEIMPROVEDIMAGELISTSUPPORT			34
// fingerprint
#define DISPID_FRM_APPID														500
#define DISPID_FRM_APPNAME													501
#define DISPID_FRM_APPSHORTNAME											502
#define DISPID_FRM_BUILD														503
#define DISPID_FRM_CHARSET													504
#define DISPID_FRM_ISRELEASE												505
#define DISPID_FRM_PROGRAMMER												506
#define DISPID_FRM_TESTER														507
// methods
#define DISPID_FRM_ABOUT														DISPID_ABOUTBOX
#define DISPID_FRM_CLICK														29
#define DISPID_FRM_GETCONTROLSTATE									30
#define DISPID_FRM_LOADSETTINGSFROMFILE							31
#define DISPID_FRM_REFRESH													DISPID_REFRESH
#define DISPID_FRM_SAVESETTINGSTOFILE								32
#define DISPID_FRM_FINISHOLEDRAGDROP								35


// _IFrameEvents
// methods
#define DISPID_FRME_CLICK														DISPID_VALUE
#define DISPID_FRME_CONTEXTMENU											1
#define DISPID_FRME_DBLCLICK												2
#define DISPID_FRME_DESTROYEDCONTROLWINDOW					3
#define DISPID_FRME_MCLICK													4
#define DISPID_FRME_MDBLCLICK												5
#define DISPID_FRME_MOUSEDOWN												6
#define DISPID_FRME_MOUSEENTER											7
#define DISPID_FRME_MOUSEHOVER											8
#define DISPID_FRME_MOUSELEAVE											9
#define DISPID_FRME_MOUSEMOVE												10
#define DISPID_FRME_MOUSEUP													11
#define DISPID_FRME_OLEDRAGDROP											12
#define DISPID_FRME_OLEDRAGENTER										13
#define DISPID_FRME_OLEDRAGLEAVE										14
#define DISPID_FRME_OLEDRAGMOUSEMOVE								15
#define DISPID_FRME_OWNERDRAW												16
#define DISPID_FRME_RCLICK													17
#define DISPID_FRME_RDBLCLICK												18
#define DISPID_FRME_RECREATEDCONTROLWINDOW					19
#define DISPID_FRME_RESIZEDCONTROLWINDOW						20
#define DISPID_FRME_XCLICK													21
#define DISPID_FRME_XDBLCLICK												22


// IOptionButton
// properties
#define DISPID_OPTBTN_APPEARANCE										1
//#define DISPID_OPTBTN_AUTOTOGGLEOPTION							2
#define DISPID_OPTBTN_BACKCOLOR											3
#define DISPID_OPTBTN_BORDERSTYLE										4
#define DISPID_OPTBTN_CONTENTTYPE										5
#define DISPID_OPTBTN_DISABLEDEVENTS								6
#define DISPID_OPTBTN_DONTREDRAW										7
#define DISPID_OPTBTN_ENABLED												DISPID_ENABLED
#define DISPID_OPTBTN_FONT													8
#define DISPID_OPTBTN_FORECOLOR											9
#define DISPID_OPTBTN_HALIGNMENT										10
#define DISPID_OPTBTN_HIMAGELIST										11
#define DISPID_OPTBTN_HOVERTIME											12
#define DISPID_OPTBTN_HWND													13
#define DISPID_OPTBTN_ICONALIGNMENT									14
#define DISPID_OPTBTN_ICONMARGINBOTTOM							15
#define DISPID_OPTBTN_ICONMARGINLEFT								16
#define DISPID_OPTBTN_ICONMARGINRIGHT								17
#define DISPID_OPTBTN_ICONMARGINTOP									18
#define DISPID_OPTBTN_IMAGE													19
#define DISPID_OPTBTN_MOUSEICON											20
#define DISPID_OPTBTN_MOUSEPOINTER									21
#define DISPID_OPTBTN_MULTILINE											22
#define DISPID_OPTBTN_OPTIONMARKONRIGHT							23
#define DISPID_OPTBTN_PROCESSCONTEXTMENUKEYS				24
#define DISPID_OPTBTN_PUSHED												25
#define DISPID_OPTBTN_PUSHLIKE											26
#define DISPID_OPTBTN_REGISTERFOROLEDRAGDROP				27
#define DISPID_OPTBTN_RIGHTTOLEFT										28
#define DISPID_OPTBTN_SELECTED											DISPID_VALUE
#define DISPID_OPTBTN_STYLE													29
#define DISPID_OPTBTN_SUPPORTOLEDRAGIMAGES					30
#define DISPID_OPTBTN_TEXT													31
#define DISPID_OPTBTN_TEXTMARGINBOTTOM							32
#define DISPID_OPTBTN_TEXTMARGINLEFT								33
#define DISPID_OPTBTN_TEXTMARGINRIGHT								34
#define DISPID_OPTBTN_TEXTMARGINTOP									35
#define DISPID_OPTBTN_USESYSTEMFONT									36
#define DISPID_OPTBTN_VALIGNMENT										37
#define DISPID_OPTBTN_VERSION												38
#define DISPID_OPTBTN_ICONINDEX											44
#define DISPID_OPTBTN_USEIMPROVEDIMAGELISTSUPPORT		45
// fingerprint
#define DISPID_OPTBTN_APPID													500
#define DISPID_OPTBTN_APPNAME												501
#define DISPID_OPTBTN_APPSHORTNAME									502
#define DISPID_OPTBTN_BUILD													503
#define DISPID_OPTBTN_CHARSET												504
#define DISPID_OPTBTN_ISRELEASE											505
#define DISPID_OPTBTN_PROGRAMMER										506
#define DISPID_OPTBTN_TESTER												507
// methods
#define DISPID_OPTBTN_ABOUT													DISPID_ABOUTBOX
#define DISPID_OPTBTN_CLICK													39
#define DISPID_OPTBTN_GETCONTROLSTATE								40
#define DISPID_OPTBTN_LOADSETTINGSFROMFILE					41
#define DISPID_OPTBTN_REFRESH												DISPID_REFRESH
#define DISPID_OPTBTN_SAVESETTINGSTOFILE						42
#define DISPID_OPTBTN_GETIDEALSIZE									43
#define DISPID_OPTBTN_FINISHOLEDRAGDROP							46


// _IOptionButtonEvents
// methods
#define DISPID_OPTBTNE_CLICK												1
#define DISPID_OPTBTNE_CONTEXTMENU									2
#define DISPID_OPTBTNE_CUSTOMDRAW										3
#define DISPID_OPTBTNE_DBLCLICK											4
#define DISPID_OPTBTNE_DESTROYEDCONTROLWINDOW				5
#define DISPID_OPTBTNE_KEYDOWN											6
#define DISPID_OPTBTNE_KEYPRESS											7
#define DISPID_OPTBTNE_KEYUP												8
#define DISPID_OPTBTNE_MCLICK												9
#define DISPID_OPTBTNE_MDBLCLICK										10
#define DISPID_OPTBTNE_MOUSEDOWN										11
#define DISPID_OPTBTNE_MOUSEENTER										12
#define DISPID_OPTBTNE_MOUSEHOVER										13
#define DISPID_OPTBTNE_MOUSELEAVE										14
#define DISPID_OPTBTNE_MOUSEMOVE										15
#define DISPID_OPTBTNE_MOUSEUP											16
#define DISPID_OPTBTNE_OLEDRAGDROP									17
#define DISPID_OPTBTNE_OLEDRAGENTER									18
#define DISPID_OPTBTNE_OLEDRAGLEAVE									19
#define DISPID_OPTBTNE_OLEDRAGMOUSEMOVE							20
#define DISPID_OPTBTNE_OWNERDRAW										21
#define DISPID_OPTBTNE_RCLICK												22
#define DISPID_OPTBTNE_RDBLCLICK										23
#define DISPID_OPTBTNE_RECREATEDCONTROLWINDOW				24
#define DISPID_OPTBTNE_RESIZEDCONTROLWINDOW					25
#define DISPID_OPTBTNE_SELECTIONCHANGED							DISPID_VALUE
#define DISPID_OPTBTNE_XCLICK												26
#define DISPID_OPTBTNE_XDBLCLICK										27


// IOLEDataObject
// properties
// methods
#define DISPID_ODO_CLEAR														1
#define DISPID_ODO_GETCANONICALFORMAT								2
#define DISPID_ODO_GETDATA													3
#define DISPID_ODO_GETFORMAT												4
#define DISPID_ODO_SETDATA													5
#define DISPID_ODO_GETDROPDESCRIPTION								6
#define DISPID_ODO_SETDROPDESCRIPTION								7
